src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,432 @@
+                }
+                else
+                {
+                    // We can't know that what is currently read is the end of the header value, that's only the case if this is the final block
+                    // If we're not in the final block, then consume nothing
+                    if (!isFinalBlock)
+                    {
+                        // Don't buffer indefinately
+                        if (span.Length > DefaultHeadersLengthLimit)
+                        {
+                            throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                        }
+                        return;
+                    }
+
+                    line = span;
+                    span = default;
+                    consumed += line.Length;
+                }
+
+                if (line.Length == 0) // an empty line means it's the end of the headers
+                {
+                    break;
+                }
+
+                colon = line.IndexOf(colonDelimiter);
+
+                if (colon == -1)
+                {
+                    throw new InvalidDataException($"Invalid header line: {line.ToString()}");
+                }
+                else
+                {
+                    if (line.Length > HeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                    }
+                    key = line.Slice(0, colon);
+                    value = line.Slice(colon + colonDelimiter.Length);
+                }
+
+                var decodedKey = GetDecodedString(key);
+                var decodedValue = GetDecodedString(value).Trim();
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+            }
+        }
+
+        private byte[] ReadBody(ref ReadOnlySequence<byte> buffer, bool isFinalBlock)
+        {
+            var sequenceReader = new SequenceReader<byte>(buffer);
+            ReadOnlySequence<byte> body;
+
+            var consumed = sequenceReader.Position;
+            var consumedBytes = default(long);
+
+            if (!sequenceReader.TryReadTo(out body, _boundary.BoundaryBytes))
+            {
+                if (!isFinalBlock)
+                {
+                    // Don't buffer indefinately
+                    if ((sequenceReader.Consumed - consumedBytes) > BodyLengthLimit)
+                    {
+                        throw new InvalidDataException($"Multipart body length limit {BodyLengthLimit} exceeded.");
+                    }
+                }
+
+                //// This must be the final key=value pair
+                //body = buffer.Slice(sequenceReader.Position);
+                //sequenceReader.Advance(body.Length);
+            }
+            buffer = buffer.Slice(sequenceReader.Position);
+            return GetBytesFromReadOnlySequence(body);
```

Comment:
Not a good idea to allocate a single byte[] for the body. These uploads could be GB in size. This should be some kind of stream.

Reply:
Yeah, I can see that now - any suggestion on how to tackle it? because I can't get a stream from a `ReadOnlySequence<>` as far as I know. Would leaving it as a `ReadOnlySequence<>` and maybe creating a stream that acts as an adapter over it work, instead of using a MemoryStream like I am using now?

Reply:
Even the sequence isn't enough, you need a stream that wraps the entire pipe and reads until the boundary is found.